File: prism/diff.py
"""Differential testing: two SCALAR functions, same byte inputs, disagree = FAILED."""

from __future__ import annotations

from typing import Any

from pathlib import Path
import os
import shutil
import struct
import subprocess
import tempfile

from prism import laws
from prism.contracts import parse_comments
from prism.fuzz import C_TYPE_SIZE, param_nbytes
from prism.models import Finding, FunctionInfo


def run_diff(functions: list[FunctionInfo], root: Path) -> list[Finding]:
    """Pair `*_a`/`*_b` (names or files) or `// diff: othername`; run the same bytes."""
    root = Path(root)
    pairs = _pairs(functions)
    if not pairs:
        return []
    out: list[Finding] = []
    for a, b in pairs:
        out.append(_diff_pair(a, b, root))
    return out


def _stem(fn: FunctionInfo) -> str:
    return Path(fn.file).stem if fn.file else ""


def _pairs(functions: list[FunctionInfo]) -> list[tuple[FunctionInfo, FunctionInfo]]:
    by_name: dict[str, list[FunctionInfo]] = {}
    for fn in functions:
        by_name.setdefault(fn.name, []).append(fn)
    used: set[int] = set()
    pairs: list[tuple[FunctionInfo, FunctionInfo]] = []

    def mark(a: FunctionInfo, b: FunctionInfo) -> None:
        used.add(id(a))
        used.add(id(b))
        pairs.append((a, b))

    # Function names foo_a / foo_b
    names = {fn.name: fn for fn in functions}
    for fn in functions:
        if id(fn) in used:
            continue
        if fn.name.endswith("_a"):
            other = names.get(fn.name[:-2] + "_b")
            if other is not None and id(other) not in used:
                mark(fn, other)

    # Header `// diff: othername`
    for fn in functions:
        if id(fn) in used:
            continue
        other_name = parse_comments(fn).get("diff")
        if not other_name:
            continue
        other = names.get(str(other_name))
        if other is not None and id(other) not in used:
            mark(fn, other)

    # Same function name in files *_a.c / *_b.c
    for group in by_name.values():
        a_fns = [f for f in group if id(f) not in used and _stem(f).endswith("_a")]
        b_fns = [f for f in group if id(f) not in used and _stem(f).endswith("_b")]
        for fa, fb in zip(a_fns, b_fns):
            mark(fa, fb)

    return pairs


def _diff_pair(a: FunctionInfo, b: FunctionInfo, root: Path) -> Finding:
    base: dict[str, Any] = dict(
        stage="diff", file=a.file, function=f"{a.name}/{b.name}",
        line=a.line, strength=laws.STRENGTH_FINDS,
    )
    if a.kind != "SCALAR" or b.kind != "SCALAR":
        return Finding(
            **base, status=laws.NEEDS_HARNESS, cls="",
            message=(
                f"differential testing needs two SCALAR functions, got {a.kind}/{b.kind}; "
                "POINTER/OTHER would invent a buffer or object"
            ),
        )
    if [(t, n) for t, n in a.params] != [(t, n) for t, n in b.params]:
        return Finding(
            **base, status=laws.ERROR, cls="",
            message="parameter lists differ; same bytes would not mean the same arguments",
        )

    src = _emit_program(a, b)
    cc = shutil.which("gcc") or shutil.which("clang")
    if not cc:
        return Finding(**base, status=laws.NOTRUN, cls="", message="no gcc/clang on PATH",
                       extra={"install": "install gcc or clang"})

    nbytes = param_nbytes(a.params)
    with tempfile.TemporaryDirectory(prefix="prism_diff_") as td:
        td_path = Path(td)
        harness = td_path / "diff.c"
        exe = td_path / "diff.exe"
        harness.write_text(src, encoding="utf-8")
        ok, err = _compile(cc, harness, exe)
        if not ok:
            return Finding(**base, status=laws.ERROR, cls="", message=f"diff compile: {err[:400]}")
        timed_out = False
        for data in _inputs(nbytes):
            st, detail = _run(exe, data)
            if st == "disagree":
                return Finding(
                    **base, status=laws.FAILED, cls="FUNC-CONTRACT",
                    message=f"{a.name} and {b.name} disagree",
                    evidence=detail[:800],
                    counterexample=data.hex(),
                    extra={"a": a.file, "b": b.file},
                )
            if st == "crash":
                return Finding(
                    **base, status=laws.CRASH, cls="FUZZ-CRASH",
                    message=f"diff harness crashed on {data[:16].hex()}",
                    evidence=detail[:800],
                    counterexample=data.hex(),
                )
            if st == "timeout":
                timed_out = True
        if timed_out:
            return Finding(
                **base, status=laws.TIMEOUT, cls="",
                message="diff harness timed out (not agreement, not a proof)",
                extra={"a": a.file, "b": b.file},
            )
    return Finding(
        **base, status=laws.CLEAN, cls="",
        message="no disagreement on sampled inputs (not a proof)",
        extra={"a": a.file, "b": b.file},
    )


def _emit_program(a: FunctionInfo, b: FunctionInfo) -> str:
    args = ", ".join(f"{t} {n}" if t else f"int {n}" for t, n in a.params)
    call = ", ".join(n for _, n in a.params)
    reads = []
    decls = []
    off = 0
    for typ, name in a.params:
        key = " ".join(typ.split()) or "int"
        sz = C_TYPE_SIZE.get(key, 4)
        decls.append(f"    {key} {name};")
        reads.append(f"    memcpy(&{name}, buf + {off}, {sz});")
        off += sz
    nbytes = off or 1
    ret_a = a.return_type.replace("*", "").strip() or "int"
    ret_b = b.return_type.replace("*", "").strip() or "int"
    return f'''#include <stdint.h>
#include <stdio.h>
#include <string.h>

static {ret_a} impl_a({args}) {{
{a.body}
}}
static {ret_b} impl_b({args}) {{
{b.body}
}}

int main(void) {{
    unsigned char buf[{nbytes}];
    if (fread(buf, 1, {nbytes}, stdin) != {nbytes}) return 0;
{chr(10).join(decls)}
{chr(10).join(reads)}
    {ret_a} ra = impl_a({call});
    {ret_b} rb = impl_b({call});
    if (ra != rb) {{
        fprintf(stderr, "DIFF %d %d\\n", (int)ra, (int)rb);
        return 2;
    }}
    return 0;
}}
'''


def _compile(cc: str, src: Path, exe: Path) -> tuple[bool, str]:
    cmd = [cc, "-O0", "-g", "-std=c11", str(src), "-o", str(exe)]
    try:
        p = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
    except subprocess.TimeoutExpired:
        return False, "compile timeout"
    if p.returncode != 0:
        return False, (p.stderr or p.stdout)[-1500:]
    return True, ""


def _run(exe: Path, data: bytes, timeout: float = 1.0) -> tuple[str, str]:
    try:
        p = subprocess.run([str(exe)], input=data, capture_output=True, timeout=timeout)
    except subprocess.TimeoutExpired:
        return "timeout", ""
    err = (p.stderr or b"").decode("utf-8", "replace")
    if p.returncode == 2 or err.startswith("DIFF"):
        return "disagree", err[-800:]
    if p.returncode < 0:
        return "crash", f"signal {-p.returncode}"
    return "ok", err[-200:]


def _inputs(nbytes: int) -> list[bytes]:
    interesting = [
        0, 1, -1, 2, 3, 42, 99, 100, 127, 128, 255,
        0x7FFFFFFF, -0x80000000, 123456, -99,
    ]
    out: list[bytes] = []
    for v in interesting:
        raw = struct.pack("<I", v & 0xFFFFFFFF) if nbytes >= 4 else bytes([v & 0xFF])
        out.append(raw[:nbytes].ljust(nbytes, b"\x00"))
    out.append(os.urandom(nbytes))
    return out

File: src/prism/stages/diff.cpp
// Stage diff: differential testing of same-named function pairs built with gcc/clang.
#include "common.hpp"

namespace prism {
namespace fs = std::filesystem;
using namespace stages_detail;

namespace {
int c_type_nbytes_key(std::string typ) {
    typ = strip(typ);
    auto it = kCTypeSize.find(typ);
    return it == kCTypeSize.end() ? 4 : it->second;
}

std::string stem_of(const FunctionInfo& fn) {
    return fs::path(fn.file).stem().string();
}

std::vector<std::pair<FunctionInfo, FunctionInfo>> diff_pairs(const std::vector<FunctionInfo>& functions) {
    std::map<std::string, std::vector<FunctionInfo>> by_name;
    for (auto& fn : functions) by_name[fn.name].push_back(fn);
    std::set<const FunctionInfo*> used;
    std::vector<std::pair<FunctionInfo, FunctionInfo>> pairs;
    auto mark = [&](const FunctionInfo& a, const FunctionInfo& b) {
        used.insert(&a);
        used.insert(&b);
        pairs.emplace_back(a, b);
    };
    std::map<std::string, const FunctionInfo*> names;
    for (auto& fn : functions) names[fn.name] = &fn;
    for (auto& fn : functions) {
        if (used.contains(&fn)) continue;
        if (fn.name.ends_with("_a")) {
            auto other = names.find(fn.name.substr(0, fn.name.size() - 2) + "_b");
            if (other != names.end() && !used.contains(other->second)) mark(fn, *other->second);
        }
    }
    for (auto& fn : functions) {
        if (used.contains(&fn)) continue;
        auto d = parse_comments(fn).diff;
        if (!d) continue;
        auto other = names.find(*d);
        if (other != names.end() && !used.contains(other->second)) mark(fn, *other->second);
    }
    for (auto& [nm, group] : by_name) {
        std::vector<const FunctionInfo*> a_fns, b_fns;
        for (auto& f : group) {
            if (used.contains(&f)) continue;
            if (stem_of(f).ends_with("_a")) a_fns.push_back(&f);
            if (stem_of(f).ends_with("_b")) b_fns.push_back(&f);
        }
        for (std::size_t i = 0; i < a_fns.size() && i < b_fns.size(); ++i) mark(*a_fns[i], *b_fns[i]);
    }
    return pairs;
}

std::string emit_diff_program(const FunctionInfo& a, const FunctionInfo& b) {
    std::string args, call;
    for (std::size_t i = 0; i < a.params.size(); ++i) {
        if (i) {
            args += ", ";
            call += ", ";
        }
        auto t = a.params[i].first;
        auto n = a.params[i].second;
        args += (t.empty() ? "int " : t + " ") + n;
        call += n;
    }
    std::string decls, reads;
    int off = 0;
    for (auto& [typ, name] : a.params) {
        auto key = strip(typ);
        if (key.empty()) key = "int";
        int sz = c_type_nbytes_key(key);
        decls += "    " + key + " " + name + ";\n";
        reads += "    memcpy(&" + name + ", buf + " + std::to_string(off) + ", " + std::to_string(sz) + ");\n";
        off += sz;
    }
    int nbytes = off ? off : 1;
    auto strip_star = [](std::string s) {
        s.erase(std::remove(s.begin(), s.end(), '*'), s.end());
        s = strip(s);
        return s.empty() ? "int" : s;
    };
    auto ret_a = strip_star(a.return_type);
    auto ret_b = strip_star(b.return_type);
    return "#include <stdint.h>\n#include <stdio.h>\n#include <string.h>\n\nstatic " + ret_a + " impl_a(" +
           args + ") {\n" + a.body + "\n}\nstatic " + ret_b + " impl_b(" + args + ") {\n" + b.body +
           "\n}\n\nint main(void) {\n    unsigned char buf[" + std::to_string(nbytes) +
           "];\n    if (fread(buf, 1, " + std::to_string(nbytes) + ", stdin) != " + std::to_string(nbytes) +
           ") return 0;\n" + decls + reads + "    " + ret_a + " ra = impl_a(" + call + ");\n    " + ret_b +
           " rb = impl_b(" + call +
           ");\n    if (ra != rb) {\n        fprintf(stderr, \"DIFF %d %d\\n\", (int)ra, (int)rb);\n        "
           "return 2;\n    }\n    return 0;\n}\n";
}

std::vector<std::vector<uint8_t>> diff_inputs(int nbytes) {
    std::vector<int> interesting{0,  1,  -1, 2,         3,          42, 99, 100,
                                 127, 128, 255, 0x7FFFFFFF, static_cast<int>(0x80000000), 123456, -99};
    std::vector<std::vector<uint8_t>> out;
    for (int v : interesting) {
        std::vector<uint8_t> raw;
        if (nbytes >= 4) {
            uint32_t u = static_cast<uint32_t>(v);
            raw = {static_cast<uint8_t>(u), static_cast<uint8_t>(u >> 8), static_cast<uint8_t>(u >> 16),
                   static_cast<uint8_t>(u >> 24)};
        } else {
            raw = {static_cast<uint8_t>(v & 0xFF)};
        }
        if (static_cast<int>(raw.size()) > nbytes) raw.resize(static_cast<std::size_t>(nbytes));
        while (static_cast<int>(raw.size()) < nbytes) raw.push_back(0);
        out.push_back(std::move(raw));
    }
    std::vector<uint8_t> rnd(static_cast<std::size_t>(nbytes));
    std::mt19937 rng{std::random_device{}()};
    for (auto& b : rnd) b = static_cast<uint8_t>(rng());
    out.push_back(std::move(rnd));
    return out;
}

Finding diff_pair(const FunctionInfo& a, const FunctionInfo& b, const fs::path&) {
    auto base = make_find("diff", laws::CLEAN, a, "", "", laws::STRENGTH_FINDS);
    base.function = a.name + "/" + b.name;
    if (a.kind != "SCALAR" || b.kind != "SCALAR") {
        base.status = std::string(laws::NEEDS_HARNESS);
        base.message = "differential testing needs two SCALAR functions, got " + a.kind + "/" + b.kind +
                       "; POINTER/OTHER would invent a buffer or object";
        return base;
    }
    if (a.params != b.params) {
        base.status = std::string(laws::ERROR);
        base.message = "parameter lists differ; same bytes would not mean the same arguments";
        return base;
    }
    if (!sandbox::allowed()) {
        // Law 9: the harness compiles and runs both scanned functions.
        auto f = sandbox::exec_notrun("diff", "diff " + a.name + "/" + b.name);
        f.file = a.file;
        f.function = a.name + "/" + b.name;
        f.line = a.line;
        return f;
    }
    auto cc = Config{}.which({"gcc", "clang"});
    if (!cc) {
        base.status = std::string(laws::NOTRUN);
        base.message = "no gcc/clang on PATH";
        base.extra["install"] = "install gcc or clang";
        return base;
    }
    auto src = emit_diff_program(a, b);
    auto td = fs::temp_directory_path() / ("prism_diff_" + std::to_string(std::random_device{}()));
    fs::create_directories(td);
    auto harness = td / "diff.c";
    auto exe = td / "diff.exe";
    {
        std::ofstream out(harness);
        out << src;
    }
    auto cr = run_argv({cc->string(), "-O0", "-g", "-std=c11", harness.string(), "-o", exe.string()}, {}, 30.0);
    if (cr.timeout || cr.rc != 0) {
        fs::remove_all(td);
        base.status = std::string(laws::ERROR);
        auto err = cr.err.empty() ? cr.out : cr.err;
        if (err.size() > 400) err.resize(400);
        base.message = "diff compile: " + err;
        return base;
    }
    int nbytes = param_nbytes(a.params);
    bool timed_out = false;
    for (auto& data : diff_inputs(nbytes)) {
        std::string in(data.begin(), data.end());
        auto rr = run_argv(sandbox::wrap_argv({exe.string()}, td), in, 1.0, sandbox::limits_for(1.0));
        bool disagree = rr.rc == 2 || rr.err.starts_with("DIFF");
        if (disagree) {
            fs::remove_all(td);
            base.status = std::string(laws::FAILED);
            base.cls = "FUNC-CONTRACT";
            base.message = a.name + " and " + b.name + " disagree";
            base.evidence = rr.err.substr(0, 800);
            std::string hex;
            for (auto b : data) {
                char buf[8];
                std::snprintf(buf, sizeof buf, "%02x", b);
                hex += buf;
            }
            base.counterexample = hex;
            base.extra["a"] = a.file;
            base.extra["b"] = b.file;
            return base;
        }
        if (rr.crashed || rr.rc < 0) {
            fs::remove_all(td);
            base.status = std::string(laws::CRASH);
            base.cls = "FUZZ-CRASH";
            std::string hex;
            for (std::size_t i = 0; i < data.size() && i < 16; ++i) {
                char buf[8];
                std::snprintf(buf, sizeof buf, "%02x", data[i]);
                hex += buf;
            }
            base.message = "diff harness crashed on " + hex;
            base.evidence = rr.err.substr(0, 800);
            std::string full;
            for (auto b : data) {
                char buf[8];
                std::snprintf(buf, sizeof buf, "%02x", b);
                full += buf;
            }
            base.counterexample = full;
            return base;
        }
        if (rr.timeout) timed_out = true;
    }
    fs::remove_all(td);
    if (timed_out) {
        base.status = std::string(laws::TIMEOUT);
        base.message = "diff harness timed out (not agreement, not a proof)";
        base.extra["a"] = a.file;
        base.extra["b"] = b.file;
        return base;
    }
    base.status = std::string(laws::CLEAN);
    base.message = "no disagreement on sampled inputs (not a proof)";
    base.extra["a"] = a.file;
    base.extra["b"] = b.file;
    return base;
}

}  // namespace

std::vector<Finding> run_diff(const std::vector<FunctionInfo>& functions, const fs::path& root) {
    auto pairs = diff_pairs(functions);
    if (pairs.empty()) return {};
    std::vector<Finding> out;
    for (auto& [a, b] : pairs) out.push_back(diff_pair(a, b, root));
    return out;
}

}  // namespace prism
